pyteal/ast/frame.py

```diff
@@ -0,0 +1,192 @@
+from typing import TYPE_CHECKING
+
+from pyteal.ast.expr import Expr
+from pyteal.types import TealType, require_type
+from pyteal.errors import TealInputError, verifyProgramVersion
+from pyteal.ir import TealBlock, TealSimpleBlock, TealOp, Op
+
+if TYPE_CHECKING:
+    from pyteal.compiler import CompileOptions
+
+
+class Proto(Expr):
+    def __init__(self, num_args: int, num_returns: int):
+        super().__init__()
+        if num_args < 0:
+            raise TealInputError(
+                f"the number of arguments provided to Proto must be >= 0 but {num_args=}"
+            )
+        if num_returns < 0:
+            raise TealInputError(
+                f"the number of return values provided to Proto must be >= 0 but {num_returns=}"
+            )
+        self.num_args = num_args
+        self.num_returns = num_returns
+
+    def __teal__(self, options: "CompileOptions") -> tuple[TealBlock, TealSimpleBlock]:
+        verifyProgramVersion(
+            Op.proto.min_version,
+            options.version,
+            "Program version too low to use op proto",
+        )
+        op = TealOp(self, Op.proto, self.num_args, self.num_returns)
+        return TealBlock.FromOp(options, op)
+
+    def __str__(self) -> str:
+        return f"(proto: num_args = {self.num_args}, num_returns = {self.num_returns})"
+
+    def type_of(self) -> TealType:
+        return TealType.none
+
+    def has_return(self) -> bool:
+        return False
+
+
+Proto.__module__ = "pyteal"
+
+
+class FrameDig(Expr):
+    def __init__(self, depth: int):
+        super().__init__()
+        self.depth = depth
+
+    def __teal__(self, options: "CompileOptions") -> tuple[TealBlock, TealSimpleBlock]:
+        verifyProgramVersion(
+            Op.frame_dig.min_version,
+            options.version,
+            "Program version too low to use op frame_dig",
+        )
+        op = TealOp(self, Op.frame_dig, self.depth)
+        return TealBlock.FromOp(options, op)
+
+    def __str__(self) -> str:
+        return f"(frame_dig: dig_depth = {self.depth})"
+
+    def type_of(self) -> TealType:
+        return TealType.anytype
+
+    def has_return(self) -> bool:
+        return False
+
+
+FrameDig.__module__ = "pyteal"
+
+
+class FrameBury(Expr):
+    def __init__(self, what: Expr, depth: int):
```

Comment:
Nit: In my brief perusal of our codebase, I've found one other example of `what` used to refer to a parameter of type `Expr`, and that only appears in a test script. On the other hand, the most common such variable name appears to be `value` and so I recommend reconsider renaming `what` to `value`.

Reply:
yes good point! This is my sinful past in writing cpp, and there are a lot of stuffs named in `what` 😅